cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -747,7 +766,32 @@ predicate readStep(Node node1, Content c, Node node2) {
  * Holds if values stored inside content `c` are cleared at node `n`.
  */
 predicate clearsContent(Node n, Content c) {
-  none() // stub implementation
+  n =
+    any(PostUpdateNode pun, Content d | d.impliesClearOf(c) and storeStepImpl(_, d, pun, true) | pun)
+        .getPreUpdateNode() and
+  (
+    // The crement operations and pointer addition and subtraction self-assign. We do not
+    // want to clear the contents if it is indirectly pointed at by any of these operations,
+    // as part of the contents might still be accessible afterwards. If there is no such
+    // indirection clearing the contents is safe.
+    not exists(Operand op, Cpp::Operation p |
+      n.(IndirectOperand).hasOperandAndIndirectionIndex(op, _) and
+      (
+        p instanceof Cpp::AssignPointerAddExpr or
+        p instanceof Cpp::AssignPointerSubExpr or
+        p instanceof Cpp::CrementOperation
+      )
+    |
+      p.getAnOperand() = op.getUse().getAst()
+    )
+    or
+    not exists(PostUpdateNode pun, Content d |
+      d.impliesClearOf(c) and
+      storeStepImpl(_, d, pun, true) and
+      c.getIndirectionIndex() > d.getIndirectionIndex() and
+      pun.getPreUpdateNode() = n
+    )
```

Comment:
AFAICT, the comment added here refers only to the first disjunct, right? It would be great to have a comment on what the role of this disjunct is.

I haven't given it much thought yet (and I probably need some lunch before I attempt to do so!), but the `n = any(...)` conjunct at the top desugards to something like:
```ql
exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  n = pun.getPreUpdateNode()
)
```
and this is _almost_ what this last disjunct is saying, except for the `c.getIndirectionIndex() > d.getIndirectionIndex()` condition, and I thought this would've been covered by the use of `>=` in the `impliesClearOf` implementation in `FieldContent`?

Reply:
> AFAICT, the comment added here refers only to the first disjunct, right? It would be great to have a comment on what the role of this disjunct is.

It refers to the combination of the two.

> I thought this would've been covered by the use of `>=` in the `impliesClearOf` implementation in `FieldContent`?

Note the `not` in front of the `exists`.

Reply:
> Note the `not` in front of the `exists`.

Right. Sorry, I should've been more clear. My point is that I'm surprised by the entire last negation isn't implied by the first conjunct. My reasoning is as follows:
If we ignore the middle condition (i.e., the one that refers to the various increment/decrements), then the predicate looks like:
```ql
n = any(PostUpdateNode pun, Content d | d.impliesClearOf(c) and storeStepImpl(_, d, pun, true) | pun).getPreUpdateNode() and
and
not exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  c.getIndirectionIndex() > d.getIndirectionIndex() and
  pun.getPreUpdateNode() = n
)
```
rewirting `n = any(...)` to an `exists` this becomes:
```ql
exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  n = pun.getPreUpdateNode()
) and
not exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  c.getIndirectionIndex() > d.getIndirectionIndex() and
  pun.getPreUpdateNode() = n
)
```
and since most of the conditions in the negation are already present outside the negation we can fold this into:
```ql
exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  n = pun.getPreUpdateNode() and
  not c.getIndirectionIndex() > d.getIndirectionIndex() and
)
```
and `not n > m` should be equivalent to `n <= m`, so this can be rewritten to:
```ql
exists(PostUpdateNode pun, Content d |
  d.impliesClearOf(c) and
  storeStepImpl(_, d, pun, true) and
  n = pun.getPreUpdateNode() and
  c.getIndirectionIndex() <= d.getIndirectionIndex()
)
```
and this last condition (i.e., the `c.getIndirectionIndex() <= d.getIndirectionIndex()` part) should already be implied by the `d.impliesClearOf(c)`, I think?

Reply:
> should already be implied by the d.impliesClearOf(c), I think?

No, that requires `c.getIndirectionIndex() >= d.getIndirectionIndex()`, so with `>=` instead of `<=`.

Reply:
Oh, I see. So in fact, the inclusion of this negation gives us that `c.getIndirectionIndex() = d.getIndirectionIndex()`?

Reply:
That makes sense. Thanks for clearing this up!

Reply:
Probably a `forex`, but yeah that be a bit more explicit about what's going on.